Restructure application / rework graph rendering

TomDijkema opened the pull request:
Commit restructures the application and refactors a lot of code to be more optimized. Especially the rendering of the graphs has been changed to a more friendly JS approach and the amount of database queries has been perked.

Adds:
- Support for saving data per year/month and country code/ror id constraint
- Percentages of the amount of specimens related to issue flag in table

Modified:
- Structure of application, is now seperated into three main directories: processing-service, api and dashboard, each with its own Dockerfile, requirements, etc.
- Reworked functions for optimized experience, mostly database and graph related functions
- Dockerfile from processing service now generates a functional image for the AWS repository

Removes:
- All support for CSV files, except the Microchanges file
- Graph objects being created in Python
- Unused and outcommented functions of past methods

TomDijkema committed:
Restructure application / rework graph rendering

Commit restructures the application and refactors a lot of code to be more optimized. Especially the rendering of the graphs has been changed to a more friendly JS approach and the amount of database queries has been perked.

Adds:
- Support for saving data per year/month and country code/ror id constraint
- Percentages of the amount of specimens related to issue flag in table

Modified:
- Structure of application, is now seperated into three main directories: processing-service, api and dashboard, each with its own Dockerfile, requirements, etc.
- Reworked functions for optimized experience, mostly database and graph related functions
- Dockerfile from processing service now generates a functional image for the AWS repository

Removes:
- All support for CSV files, except the Microchanges file
- Graph objects being created in Python
- Unused and outcommented functions of past methods

TomDijkema committed:
Sonar / Clean-up commit

Commit resolves Sonar issues and cleans up code, adds different logging to main function, type annotation and docstrings to most functions.

TomDijkema committed:
Sonar commit 2

Second commit for resolving Sonar issues

TomDijkema committed:
Sonar commit 3

Third commit for resolving Sonar issues

TomDijkema committed:
Dockerfile

samleeflang committed:
Docker + k8s files for API

TomDijkema committed:
Sonar commit 4

Fourth commit for resolving Sonar issues

TomDijkema committed:
Merge branch 'feature-graph_endpoint' of github.com:DiSSCo/network-status-overview into feature-#33762400

TomDijkema committed:
Fix weird graph loading issues

Commit for fixing the weird visual issues when loading graphs that were displayed as none

TomDijkema committed:
Sonar commit 5

Fifth commit for fixing Sonar issues

TomDijkema committed:
Sonar commit 6

Sixth commit for solving Sonar issues

TomDijkema committed:
Feedback Tweaks

Commit adds tweaks and small additions based upon feedback

Adds:
- Total specimens colum in datasets table graph
- Calculated percentages between GBIF and GeoCASe in datasets table graph
- X-axis for both datasets bar as specimens bar graph
- Added the ROR id to an organisation as part of the dropdown (select) list for rendering graphs

Modifies:
- Somewhat better hover (tooltips_ for specimens bar graph
- Colors of lines in specimens line graph now complement general color scheme

samleeflang commented on api/draw_functions/draw_specimens.py:
```diff
@@ -0,0 +1,149 @@
+]
+
+# List of used specimen types
+basis_of_record = [
```
Deze lijst kom ik op meerdere plekken tegen, ik zou deze ergens als constant plaatsen zodat ze allemaal dezelfde lijst gebruiken

TomDijkema replied:
Created a new file called 'constants' that can be imported to use the common lists

samleeflang commented on api/draw_functions/draw_specimens.py:
```diff
@@ -0,0 +1,149 @@
+# Color schema used in the graphs
+colors = [
```
Deze lijst kom ik vaker tegen, maak hier een constant van die overal wordt gebruikt. Mocht er dan een aanpassing komen dan hoef je het maar op 1 plek aan te passen

TomDijkema replied:
Created a new file called 'constants' that can be imported to use the common lists

samleeflang commented on api/main.py:
```diff
@@ -0,0 +1,112 @@
+
+import graphical_functions
+
+app = Flask(__name__)
+CORS(app, allow_headers=['Content-Type', 'Access-Control-Allow-Origin',
+                         'Access-Control-Allow-Headers', 'Access-Control-Allow-Methods'])
+CSRFProtect(app)
+
```
We need to check these for security. CORS should work out of the box, you might only need this annotation for development. CSRF protection should by default be included in Flask not sure it is needed to add the flask_wtf as well.

TomDijkema replied:
Removed extra CSRF import, reduced CORS allow_headers to: 'Content-Type', the only that was needed to be able to run the application without immediate CORS errors

samleeflang commented on api/main.py:
```diff
@@ -0,0 +1,112 @@
+    return fig
+
+
+@app.route('/get_organisations', methods=['GET'])
+def get_organisations():
+    """ Retrieves the list of organisations from the database
+        :return: The organisations list
+    """
+
+    response = requests.get('https://sandbox.dissco.tech/api/v1/organisation/tuples')
+
+    return jsonify(response.json())
```
Don't think this call is necessary

TomDijkema replied:
Removed call and replaced url in Javascript with: https://sandbox.dissco.tech/api/v1/organisation/tuples.
However, api needs to set a CORS allow-origin header for the DiSSCo website's origin to be able to use (don't know if proxy would work like the React one, also using multiple endpoints/apis).

samleeflang replied:
Spoke about this and we will keep the Flask for redirect to avoid CORS

samleeflang commented on api/main.py:
```diff
@@ -0,0 +1,112 @@
+                      'SE': 'Sweden',
+                      'GB': 'United Kingdom'}
+
+    return jsonify(countries_list)
```
I remember now that we spoke about this and will only list countries which are in the list Wouter provided. I will implement this in the API

samleeflang replied:
API is in the air

TomDijkema replied:
Replaced list with API respone

samleeflang commented on api/query_database.py:
```diff
@@ -0,0 +1,464 @@
+    return database_config
+
+
+def prepare_country(global_data: dict, country: list, country_code: str) -> dict:
```
Still a large function. Some part could get extracted, ok for now but if there is time left might be nice too look at. Is the some for several other functions in this file

TomDijkema replied:
Ok, will keep in mind

samleeflang commented on api/requirements.txt:
```diff
@@ -1,6 +1,8 @@
-plotly==5.7.0
 requests~=2.22.0
+plotly==5.7.0
```
Can be updated to 5.9.0

TomDijkema replied:
Updated

samleeflang commented on dashboard/dashboard.css:
```diff
@@ -0,0 +1,141 @@
+.dashboardSection {
```
Still a lot of CSS which should generally be avoided. However you indicated that you need it to make it work with Wordpress so might be relevant here, maybe add a comment

TomDijkema replied:
Agree, tried to use as less CSS as possible, so mostly colors and font stuff, however, constructed the lay-out of the dashboard with custom heights, was struggling a bit with that, I think it can be optimized but I could not get it to work the first try with only Bootstrap. Will add a comment and possibly a thing to solve in future

samleeflang commented on dashboard/dashboard.js:
```diff
@@ -0,0 +1,383 @@
+function getOrganisationsList() {
+  $.ajax({
+    type: "GET",
+    url: "https://sandbox.dissco.tech/api/v1/network-overview/get_organisations",
```
See if you can switch this to the generic organisations API

TomDijkema replied:
Removed call and replaced url in Javascript with: https://sandbox.dissco.tech/api/v1/organisation/tuples.
However, api needs to set a CORS allow-origin header for the DiSSCo website's origin to be able to use (don't know if proxy would work like the React one, also using multiple endpoints/apis).

samleeflang commented on dashboard/dashboard.js:
```diff
@@ -0,0 +1,383 @@
+  /* Receive countries data */
+  $.ajax({
+    type: "GET",
+    url: "https://sandbox.dissco.tech/api/v1/network-overview/get_countries",
```
I think we should remove the 'get' part from the URL. Your type is already 'get' so '/countries' (instead of get_countries) should be enough

TomDijkema replied:
Removed 'get' part so url resolves to: '/countries', as well for graphs: '/graph'

samleeflang commented on database.ini:
```diff
@@ -1,5 +1,7 @@
-user=
-password=
+user=tdijkema
+password=MeYP8SLv8cokeYIBF0J5i8xaRLKNSikqHAG1aIjV
```
No Password in git!!! First rule of git commits ;-) Never add a password into git. It is a public repository so everyone who sees this can log into our database and add massive a mount of data (and drain our money). I will reset your password and send you a new one

TomDijkema replied:
It somehow came through my Gitignore, removed file and readded so Gitignore resets

samleeflang commented on processing_service/GeoCASe_functions.py:
```diff
@@ -2,6 +2,7 @@
 
 
 geocase_endpoint = "https://geocase.eu/api"
+sonar_constant_facet_field = 'facet.field'
```
I would not use sonar in the name. If you read this I now think we need this variable because of sonar which I don't think is the case. Just call it facet_field

TomDijkema replied:
Removed Sonar mention and renamed variable to: facet_field_constant (as well as another variant: datasets_constant)

samleeflang commented on sources/.~lock.microchanges.csv#:
```diff
@@ -0,0 +1 @@
+,tom,LT006954-ubuntu,05.07.2022 14:29,file:///home/tom/.config/libreoffice/4;
```
This doesn't have the be commited. You can add it to the gitignore

TomDijkema replied:
Added file to Gitignore

samleeflang commented on processing_service/insert_to_database.py:
```diff
@@ -0,0 +1,166 @@
+current_month = dt.now().strftime('%B')
+
+# Temporary mapping between GBIF and GeoCASe
+organisation_mapping = {
```
I have the feeling I saw this mapping list before, if so try to keep it in one place (could also be the db)

TomDijkema replied:
This list together with country_mapping is created once at the top of this file to solve the mapping between GBIF and GeoCASe because at the moment there are no common identifiers between the two

TomDijkema replied:
Definitely a temporary solution

samleeflang commented on processing_service/main.py:
```diff
@@ -0,0 +1,81 @@
+    toc = time.perf_counter()
+    passed_time = round(toc - tic, 2)
+
+    logging.info(f'Finished in {passed_time} seconds')
```
Nice to have this information although we also have the time in the log statement through which you can calculate the time it took but no problem to make it explicit

samleeflang reviewed (commented):
Tom nice work! With the limitation you had (fitting it in Wordpress) it looks good. Few general things:
- Some functions can be cut up in separate functions with explicit naming
- Found some duplication in constants (same list/mappings) which is better to define once
- The CSS and HTML in python gives me itches, normally I would have this only in the frontend but don't think we can get around this with the Wordpress setup.

TomDijkema committed:
Review commit

Commit applies changes comform the review.

Adds:
- Constants file with constant lists

Changes:
- Countries are now derived from organisation api

Removes:
- Additional CSRF import

TomDijkema committed:
Readded CSRF component

Documentation says when not using Flask forms to use the CSRF component inside the python code, component gets applied to app instance
